### ISSUE TITLE
Refactor old Distribution base class

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pymc#5308]
PyMC distribution classes are weird objects that hold RandomVariables, logp, logcdf and moment methods together (basically doing runtime dispatching) and manage most of the non-RandomVariable kwargs that users are familiar with (observed, transformed, size/dims) and behind the scenes actions like registration in the model.

This exists mostly for backwards compatibility with V3 and ease of developer refactoring, but the current result is far from pretty.

We need to figure out a more elegant/permanent architecture now that many things that existed to accommodate V3 limitations no longer hold.

## Distribution
`Distribution` is currently performing the following tasks:

https://github.com/pymc-devs/pymc/blob/75ea2a80cb27773e93b7b207043077953940e6ff/pymc/distributions/distribution.py#L135

1. Input validation:
    1. Raising `FutureWarnings` for `testval` kwarg
    1. Raising TypeError when distribution is initialized outside of a Model context
    1. Raising TypeError when name is not given to a distribution
    1. Raising ValueError when more than one of dims/shape/size is given
1. Convert alternative parametrizations to standard parametrization (e.g, `tau` -> `sigma`). This is done by the `.dist` methods.
1. Add informative attribute errors for deprecated `logp`, `logcdf`, `random` methods
1. Resize the final RV based on observed, shape, dims or size
1. Provides the `.dist()` API to create an unnamed RV that is not registered in the model. This type of variables is necessary for use in Potentials and other distribution factories that use RVs as building blocks such as Bound and Censored distributions, as well as Mixtures and Timeseries once they get refactored for V4

## DistributionMeta
In addition we have a `DistributionMeta` that does the following:

https://github.com/pymc-devs/pymc/blob/75ea2a80cb27773e93b7b207043077953940e6ff/pymc/distributions/distribution.py#L70

1. Dispatch the `logp`, `logcdf`, `moment`, `default_transform` methods defined in the old PyMC distributions to apply to the respective `rv_op`
1. Register the `rv_op` type as subclass of the old style PyMC distribution, so that V3 Discrete/Continuous subclass checks still work?

```python
isinstance(pm.Normal.dist().owner.op, pm.distributions.Continuous)  # True
```

If we want to get rid of `Distribution` we probably need to statically dispatch our methods to the respective `rv_op`. That is nothing special, and is how we do it for aeppl from the get go: https://github.com/aesara-devs/aeppl/blob/38d0c2ea4ecf8505f85317047089ab9999d2f78e/aeppl/logprob.py#L104-L130

**Comment by ricardoV94** [comment]
> > They serve the same goal the "standard" Distribution class is doing right now: basically manage all non distribution kwargs: shape/dims, transformed, observed; and behind the scenes magic things like value variables and rngs

> The base `Distribution` class was only ever supposed to make the hand-off to the `Model` class, which would then handle all the generic tasks (e.g. value variables, RNGs, shape/dims, etc.), and the rest was just for backward compatibility (e.g. the now completely unnecessary `Disribution.dist` interface).  Notice how instances of these classes are never made, and that only static/class/type-level functions are used.

> By extending `Distribution` and adding more logic to those classes, it could become considerably more difficult to unwind those temporary backward compatibility-only choices, and they'll eventually become a permanent and confusing part of v4's design.

> This isn't the time or place to start addressing all that, but these are the kinds of design considerations that need to go alongside changes/additions to the relevant areas of code (i.e. `Distribution`-related code).

_Originally posted by @brandonwillard in https://github.com/pymc-devs/pymc/issues/5169#issuecomment-1004289756_

**Comment by fonnesbeck** [comment]
Do we have a spec in mind for a replacement class? Maybe an explicit list of what needs to be removed and added with respect to the current class would be a good place to start.

**Comment by ricardoV94** [comment]
> Do we have a spec in mind for a replacement class? Maybe an explicit list of what needs to be removed and added with respect to the current class would be a good place to start.

I will update the issue tomorrow with those

**Comment by canyon289** [comment]
> to accommodate V3 limitations no longer hold.
If youre easily able to list these no longer valid limitations while writing the doc I would also appreciate that (to help with my personal understanding). No obligation if itll take you a lot of time

**Comment by ricardoV94** [comment]
Updated the top post to mention every function of `Distribution` classes that I could find. Also highlighted some of the points I think will be a bit more challenging, or may need more drastic API changes.

**Comment by fonnesbeck** [comment]
`eval()` should be aliased to something more intuitive like `sample()` or `draw()` where appropriate.

**Comment by ricardoV94** [comment]
> `eval()` should be aliased to something more intuitive like `sample()` or `draw()` where appropriate.

`eval` is just the standard Aesara debug feature that exists for any node. We can have a function wrapper that compiles the "proper" aesara function and takes a given number of draws. For analogy with V3 it could be named `pm.random`, but I like the `draw` name better.

The question of RNGs with default updates exists anyway, but could arguably be offloaded to `pymc.aesaraf.compile_pymc` which already does this for Simulator variables anyway, and which `pm.random`/ `pm.draw` would call. That would be one less thing that Distribution has to be concerned with.

https://github.com/pymc-devs/pymc/blob/75ea2a80cb27773e93b7b207043077953940e6ff/pymc/aesaraf.py#L968

**Comment by fonnesbeck** [comment]
Yeah, I like a functional approach there, and I agree `draw` is better than `sample` since the latter is already associated with MCMC sampling.

**Comment by michaelosthege** [comment]
What is the selling point of it?
It's a big big refactoring that forces us into an API breakage.
And after adding all those decorators we'd probably end up with more lines of code per distribution; considerably harder to comprehend.

-------

First of all, if we create no instances of _any_ PyMC distribution. Why do they exist at all?

We distinguish between variables that are
  * registered by name in a model: `pm.MyDistribution("name", some, params)` or
  * created as unregistered, auxiliary variables, that may still contribute to a model graph: `pm.MyDistribution.dist(some params)`

Pure Aesara RV `Op`s don't make this distinction (point 7).

Then PyMC distributions have a much richer API & behavior compared to Aesara `RandomVariable` `Op`s (points 1,2,3,5).

So __we need to wrap the `Op`__.
But should these adapters be a `class MyDistribution(...)` or a `def MyDistribution(...)`?

Neither `MyDistribution` nor `MyDistribution.dist()` seem to get much love, and generally it feels like there's a recent hype for functional API designs.
But class vs. function aside, let's consider just the resulting user API syntax:
1. `pm.Normal("name", 0, 1)` and `pm.Normal.dist(0, 1)` → the current API. Done with `class Normal(pm.Distribution)` because otherwise we'd be monkey-patching things onto functions and nobody wants that. We can still rename `.dist()` to `.cool()` or something.
2. `pm.Normal("name", 0, 1)` and `pm.Normal(0, 1)` → doesn't work as long as `name, mu, sd` etc. are positional. Same for having a kwarg like `pm.Normal(0, 1, register=False)`. In C# this would be trivial, by the way.
3. `pm.Normal("name", 0, 1)` and `pm.normal(0, 1)` → Both could be functions, but maybe also easy to confuse?

So we could move away from `class MyDistribution(Distribution)` only if we switch to syntax 3.

Then every distribution needs to get a `def mydistribution(...)` __and__ `def MyDistribution(...)`.
The latter could be a one-liner: `MyDistribution = pm.as_rich_RV(mydistribution)`.

Then we'd have to dispatch `logp` and `get_moment` etc. onto the RV Op as @ricardoV94 suggested.
The resulting code would no longer be grouped by distribution, and have a ton of dispatch decorators doing the job that's currently done by `DistributionMeta`.

-----
> 6. Taking care of default updates for the RandomState variables so that returned variables "look" random by default.

This once bit me really hard when I tried to demo Aesara to someone.
I wanted to show how conveniently one can do some symbolic calculations and `.eval()` things as random numbers.
...took me an hour to understand that I was use to a really awesome behavior that was actually a PyMC, not an Aesara feature.

If Aesara made the non-deterministic `.eval()` the default---like with PyMC RVs---it would be a lot more beginner friendly & interesting for people doing, for example Monte Carlo sampling.

**Comment by ricardoV94** [comment]
> pm.Normal("name", 0, 1) and pm.Normal(0, 1) → doesn't work as long as name, mu, sd etc. are positional. Same for having a kwarg like pm.Normal(0, 1, register=False). In C# this would be trivial, by the way.

Can't we just check if the first variable is a string and react accordingly?

**Comment by ricardoV94** [comment]
> Then we'd have to dispatch logp and get_moment etc. onto the RV Op as @ricardoV94 suggested.
The resulting code would no longer be grouped by distribution, and have a ton of dispatch decorators doing the job that's currently done by DistributionMeta.

We already do that anyways, just via the cryptic MetaDistribution. Structurally, what difference does it make if the logp and logcdf are fake methods inside a fake classes or real functions one after the other? 

Does this seem so bad?
https://github.com/aesara-devs/aeppl/blob/333108143a1f4b63d9fcd9842dbe35457025c180/aeppl/logprob.py#L104-130

More importantly users should not be calling these pseudo methods themselves because they expect as inputs the already parsed and symbolic canonical parameters of the distribution. For some (many) distributions these have nothing to do with what they would pass into `.dist`, so making those accessible from the distribution classes (via autocompletion) is actually error prone.

**Comment by mitch-at-orika** [comment]
Posting this incase it helps... I came across an issue like this one on [discourse](https://discourse.pymc.io/t/equvilant-apis-in-pymc-v4/8920) while trying to use v3 code in v4. Trying to do some of the plots in [BCB](https://bayesiancomputationbook.com/notebooks/chp_07.html#figure-7-4) the authors use FreeRV.distribution.all_trees[..].predict_output() in creating insightful plots. In v4 the error is 'TensorVariable' object has no attribute 'distribution'. 

I tried to solve but it seems in v3, model.py used def var() and returned a <>RV pymc object where distribution=dist. However in v4 we have def register_rv() which returns an aesara tensor variable? Are we breaking this functionality in v4? Sorry if it is something obvious I am missing.

**Comment by michaelosthege** [comment]
@mitch-at-orika yes, there is a breaking change w.r.t. what's returned. In most cases that's not a problem, but it looks like `BART` monkey-patched some information onto the `TensorVariable`. This is no longer possible with `v4` where these calls don't necessarily return the same tensor that was created by the underlying distribution.

@aloctavodia I believe you can link to relevant `BART` issues, or even explain how to access the `all_trees` object in `v4`?

**Comment by mitch-at-orika** [comment]
Thanks for the explanation Michael, it is reassuring it was a BART only option, I originally thought I had just missed this functionality of pymc vars until now.

**Comment by ricardoV94** [comment]
Here is some pseudo-code that might suffice?

```python
from functools import partial, wraps
import pytensor.tensor.random as ptr

from pymc.model import modelcontext
from pymc.util import UNSET

def dist_to_rv_op(dist):
    
    @wraps(dist)
    def inner_func(*args, size=None, shape=None, **kwargs):
        if shape is not None and size is not None:
            raise ValueError("Cannot pass both size and shape")
        if shape is not None:
            # Call dist without size to find out ndim_supp
            ndim_supp = dist(*args).owner.op.ndim_supp
            if ndim_supp == 0:
                size = shape
            else:
                size = shape[:-ndim_supp]
        return dist(*args, size=size, **kwargs)
    return inner_func


def dist_to_model_rv(dist):
    
    rv_op = dist_to_rv_op(dist)
    
    @wraps(dist)
    def inner_func(name, *args, dims=None, transform=UNSET, observed=None, **kwargs):
        # TODO: Handle dims
        rv = rv_op(*args, **kwargs)
        model = model_context(model)
        return model.register_rv(rv, name=name, dims=dims, transform=None, observed=observed)
    
    inner_func.dist = rv_op
    return inner_func


def normal_dist(mu=0.0, sigma=1.0, **kwargs):
    return ptr.normal(mu, sigma, **kwargs)

Normal = dist_to_model_rv(normal_dist)
```

**Comment by thomasaarholt** [comment]
@ricardoV94 I think I understand this, but just to be certain:

You have written a function that constructs the correct pytensor TensorVariable, and then you have a wrapper class that associates that variable with whatever model context manager this is created within.

Yes, I believe this should work. [Here is an example ](https://gist.github.com/thomasaarholt/5f2cf793a70aef39a6154d821ce9d05f)of how you would type `register_model_rv` in order to propogate the function signature of `normal_dist` onto `Normal`.

**Comment by thomasaarholt** [comment]
I think this introduces another potential problem, however. This new `Normal` object does not have the `.dist` method or any other methods or properties associated with it. Or perhaps that's ok, and this would be a breaking change?

**Comment by ricardoV94** [comment]
> I think this introduces another potential problem, however. This new `Normal` object does not have the `.dist` method or any other methods or properties associated with it. Or perhaps that's ok, and this would be a breaking change?

They are being monkey-patched here:

```python
    # Monkey-patch useful attributes
    if rv_type is not None:
        inner_func.rv_type = rv_type
    inner_func.dist = dist
    return inner_func
```

I am not sure that's the best approach, but the current fake classes also seem odd. Maybe what's done by `Distribution.__new__()` now should be done by `Distribution().__call__()`? It's still a pretty useless object, but it better keeps the two variations together

**Comment by thomasaarholt** [comment]
I believe we desire two interfaces per distribution that require similar (but not identical) signatures:
We want a `dist` function, and a "register the distribution with a model" function.

For the case of the normal distribution, and only using a minimal number of variables to represent the difference:
```python
def normal_dist(mu, sigma, **kwargs):
    ...
    
def Normal(name, mu, sigma, dims, observed, **kwargs):
    ...
```

I do not believe there is a way to programmatically produce a _static_ function signature for `Normal` by using the one for `normal_dist` through such mechanisms as `ParamSpec`, as discussed in [this gist](https://gist.github.com/thomasaarholt/5f2cf793a70aef39a6154d821ce9d05f?permalink_comment_id=4859804#gistcomment-4859804).

In this case `Normal` is a function, but it could easily be a class like it is today, and we could keep `.dist` as its method. I like your suggestions for a simpler inheritance structure. But wouldn't it be better to stick the contents of `__call__` inside `__init__` instead? Then rename `NormalDist` to `Normal` and call `super().__init__()` inside that classes `__init__`? Then we only refer to `Distribution` and `Normal`, rather than `Distribution`, `NormalDist` and `Normal`.

The only solution that I think makes sense is to type the signature out in both cases, using mechanisms like [`TypedDict` to keep kwargs nice](https://stackoverflow.com/a/76466821/1018861) and minimize the amount of code duplication. I'm working on a MVP in a gist to demonstrate.

**Comment by ricardoV94** [comment]
> But wouldn't it be better to stick the contents of __call__ inside __init__ instead?

init is not allowed to return anything other than None. PyMC distributions are returning completely new objects (TensorVariables), and in that sense `__new__` as done now is appropriate although an overkill because it's just a fancy function call.

**Comment by thomasaarholt** [comment]
Ah yes, that makes sense Thank you for clarifying these things for me!

**Comment by thomasaarholt** [comment]
I took a stab at rewriting the Normal class as a function (`def Normal`), and wrote it up in [this gist](https://gist.github.com/thomasaarholt/670da76d0209aee29751166261e7ada7).

**Comment by ricardoV94** [comment]
> I took a stab at rewriting the Normal class as a function (`def Normal`), and wrote it up in [this gist](https://gist.github.com/thomasaarholt/670da76d0209aee29751166261e7ada7).

Looks good. However, the whole code inside `Normal` is distribution-agnostic so it should just be some decorator in the end? Something like `Normal = [wrap_dist|model_var_from_dist](Normal_dist)?  that copies / extends the docstrings from `dist` (if that's possible)?

Also how bad would be to monkey_patch `Normal.dist = normal_dist` to keep backward compat?

**Comment by thomasaarholt** [comment]
These are good points! I haven't quite worked out a decorator approach, but I figured out a way to keep the current API while refactoring things to a "function-based" approach. `Normal` is still a class, but used quite differently. 

The key part is `Normal = NormalDistribution()`, and `Normal` is the thing that is exposed as `pm.Normal`. [Take a look](https://gist.github.com/thomasaarholt/670da76d0209aee29751166261e7ada7)!

**Comment by ricardoV94** [comment]
You need to consider observed size before you create the dist so that `pm.Normal("x", observed=np.zeroes(5))` has the right shape. 

Anyway the downside of your last approach is you need to manually duplicate the signature of the dist and call. Also it seems not much better to initialize `Normal=NormalDistribution()` since we never really use the class but I see the advantage that dist now exists transparently in it. Should be a staticmethod though since `self` isn't ever used.

**Comment by thomasaarholt** [comment]
> the downside of your last approach is you need to manually duplicate the signature of the dist and call

Assuming that our goal is to have a function signature for the `Normal` and `Normal_dist` functions that can help people correctly pass arguments to these functions, then I think there is no way around that. I don't find this particularly problematic though.

I've refactored the gist into a [small repo](https://github.com/thomasaarholt/pymc-dist-func), just to be able to spread the functions and classes into separate modules.
Here is the current version of the [normal distribution](https://github.com/thomasaarholt/pymc-dist-func/blob/main/src/pymc_dist_func/normal.py), and [here is an attempt at a decorator version](https://github.com/thomasaarholt/pymc-dist-func/blob/9e9eb751f6d4ce6d4831e2017b573c1d3b624d33/src/pymc_dist_func/normal_decorator.py#L75) - but I don't think this is a good idea. It adds a lot of boilerplate code with little gain.

**Comment by ricardoV94** [comment]
The idea of the decorator was to wrap dist, so you only have to define one of the objects (dist in this case).

We don't need classes if the only reason for them is so that Normal.dist exists. I rather just monkey patch tbh. There's nothing "class"ey about them otherwise. They don't hold state or do anything.

**Comment by thomasaarholt** [comment]
Right, but the problem with monkey patching is that we won't get autocompletion of the `Normal.dist` monkey-patched function when importing `Normal`. Within the same file, pyright **will** give you the docstring and arguments:
<img width="314" alt="Screenshot 2024-10-30 at 22 02 37" src="https://github.com/user-attachments/assets/a413deb8-745e-408f-88da-aa313b8f3a88">

But if we monkey patch in one file, and then import it in another file, then pyright **won't** tell us that the method exists. Here I define it on the left and import it in the file on the right.
<img width="1590" alt="image" src="https://github.com/user-attachments/assets/16a305b8-65ee-4ada-8b66-4373054d05cb">

We don't get autocomplete, arguments or docstring:
<img width="763" alt="Screenshot 2024-10-30 at 22 12 35" src="https://github.com/user-attachments/assets/fb3a7ed8-6083-48f4-952e-83469dc96c65">

This is why the class solution is necessary - it's the only one that will let us "attach" stuff to the object in a manner that type checkers will recognize.

**Comment by ricardoV94** [comment]
Right, but then what changed in your previous attempt? We moved `__new__` into `__call__` and duplicated the signatures between `__call__` and `__dist__`?

Not saying it's worse, just double checking what we're trading off

**Comment by thomasaarholt** [comment]
Right, I'm interpreting your question as something like "We wanted to move from a class-based approach to a function-based approach, but now we have moved from one class approach to another one, how is this better?". I think that's a very valid question.

Just to summarize, the goals I see with this github issue are:
1. Get rid of the complicated metaclass
2. Add proper arguments to the distributions, resolving #6083
3. Keep the previous API of e.g. `Distribution()` and `Distribution.dist()`.

So, in pymc today, `Normal.dist` has a [proper function signature](https://github.com/pymc-devs/pymc/blob/d37744473ce54afc3bfee5fd4e19161138595b58/pymc/distributions/continuous.py#L487), while `Normal` doesn't ([it uses `*args`](https://github.com/pymc-devs/pymc/blob/d37744473ce54afc3bfee5fd4e19161138595b58/pymc/distributions/distribution.py#L436) - because it gets intercepted by the metaclass's `__new__` method.

Using that metaclass approach, we have no way to statically (by that I mean "in a manner in which the type checker can infer without running the code") give our Distribution named arguments (e.g. `mu` and `sigma`).

Using only functions (`def normal` and `def normal_dist`) we could do everything except keep the previous API as per my previous comment above about the problems with monkey-patching.

My solution solves the problem with monkey-patching the function approach by statically defining the `dist` method on the class, so that it is well-defined at class definition time. You can think of my class-approach as more like a `API factory` for keeping the same API.

The code flow is also completely linear: The Distribution-class approach is nonlinear because it [defines `rv_op` on the subclass](https://github.com/pymc-devs/pymc/blob/d37744473ce54afc3bfee5fd4e19161138595b58/pymc/distributions/continuous.py#L484), but refers to it in the [superclass](https://github.com/pymc-devs/pymc/blob/d37744473ce54afc3bfee5fd4e19161138595b58/pymc/distributions/distribution.py#L580) and [metaclass](https://github.com/pymc-devs/pymc/blob/d37744473ce54afc3bfee5fd4e19161138595b58/pymc/distributions/distribution.py#L102).
